File: examples2d/debug_box_ball2.rs
use rapier_testbed2d::TestbedViewer;
use rapier2d::prelude::*;

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let rad = 1.0;
    let rigid_body = RigidBodyBuilder::fixed()
        .translation(Vector::new(0.0, -rad))
        .rotation(std::f32::consts::PI / 4.0);
    let collider = ColliderBuilder::cuboid(rad, rad);
    let _ = world.insert(rigid_body, collider);

    // Build the dynamic box rigid body.
    let rigid_body = RigidBodyBuilder::dynamic()
        .translation(Vector::new(0.0, 3.0 * rad))
        .can_sleep(false);
    let collider = ColliderBuilder::ball(rad);
    let _ = world.insert(rigid_body, collider);

    /*
     * Set up the testbed.
     */
    viewer.set_world(&mut world);
    viewer.look_at(Vec2::ZERO, 50.0);

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
        }
    }
    Ok(())
}

File: python/rapier-testbed/rapier_testbed/examples2/debug_box_ball2.py
"""Port of examples2d/debug_box_ball2.rs."""
from __future__ import annotations

import math

import rapier2d as rp
from .._registry import register

CATEGORY = "Debug"
NAME = "Box ball"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    rad = 1.0
    handle = bodies.insert(
        rp.RigidBody.fixed().translation((0.0, -rad)).rotation(math.pi / 4.0)
    )
    colliders.insert_with_parent(rp.Collider.cuboid(rad, rad), handle, bodies)

    handle = bodies.insert(
        rp.RigidBody.dynamic().translation((0.0, 3.0 * rad)).can_sleep(False)
    )
    colliders.insert_with_parent(rp.Collider.ball(rad), handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 0.0), zoom=50.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
